Documentation/grpc-metadata.md

```diff
@@ -0,0 +1,223 @@
+A metadata can be created using package [metadata](https://godoc.org/google.golang.org/grpc/metadata).
+The type MD is actually a map from string to a list of strings:
+
+```go
+type MD map[string][]string
+```
+
+The content of metadata can be accessed as reading normal maps.
+Note that the value type of this map is `[]string`.
+
+### Creating a new metadata
+
+A metadata can be created from a `map[string]string` using function `New`:
+
+```go
+md := metadata.New(map[string]string{"key1": "val1", "key2": "val2"})
+```
+
+Another way is to use `Pairs`.
+Values with the same key will be merged into a list:
+
+```go
+md := metadata.Pairs("key1", "val1", "key2", "val2")
+```
+
+__Note:__ all keys in metadata will be lowercased.
+
+#### Storing binary data in metadata
+
+In metadata, keys are always strings. But values can be strings or binary data.
+To store binary data value in metadata, simply add "-bin" surfix to the key.
+The values with "-bin" surfixed keys will be encoded when creating the metadata:
+
+```go
+md := metadata.Pairs(
+  "key", "string value",
+  "key-bin", string([]byte{96, 102}), // this binary data will be encoded (base64)
+)
+```
+
+__Note:__ encoded data will be decoded automatically after being transferred.
+So you don't need to call `DecodeKeyValue` explicitly when reading.
+
+### Retrieving metadata from context
+
+Metadata can be retrieved from context using `FromContext`:
+
+```go
+func (s *server) SomeRPC(ctx context.Context, in *pb.SomeRequest) (*pb.SomeResponse, err) {
+	// read metadata from the context
+	md := metadata.FromContext(ctx)
+}
+```
+
+### Updating metadata
+
+To update a metadata is to update the map underneath:
+
+```go
+md["key3"] = []string{"val3"}
+md["key3"] = append(md["key3"], "val3-2") // append "val3-2" as the second value of "key3"
+```
+
+
+## Sending and receiving metadata - client side
+
+[//]: # "TODO: uncomment next line after example source added"
+[//]: # "Real metadata sending and receiving examples are available [here](TODO:example_dir)."
+
+### Sending metadata
+
+To send metadata to server, the client can wrap the metadata into a context using `NewContext`, and make the RPC call with this context:
+
+```go
+md := metadata.Pairs("key", "val")
+
+// create a new context with this metadata
+ctx := metadata.NewContext(context.Background(), md)
+
+// make unary RPC call
+response, err := client.SomeRPC(ctx, someRequest)
+
+// or make streaming RPC
+stream, err := client.SomeStreamingRPC(ctx)
+```
+
+
+### Receiving metadata
+
+Metadata a client can receive includes header and trailer.
+
+#### Unary call
+
+Header and trailer sent along with a unary call can be retrieved using function [Header](https://godoc.org/google.golang.org/grpc#Header) and [Trailer](https://godoc.org/google.golang.org/grpc#Trailer) in [CallOption](https://godoc.org/google.golang.org/grpc#CallOption):
+
+```go
+var header, trailer metadata.MD // variable to store header and trailer
+r, err := client.SomeRPC(
+	ctx,
```

Comment:
<tab>ctx
<space><space><space><space>someRequest

switch to use all tab or all space to be consistent?

so the fields are not properly aligned.